src/macros-by-example.md

```diff
@@ -195,6 +195,220 @@ compiler knows how to expand them properly:
+}
+```
+
+`count` can not be placed inside the innermost repetition.
+
+### ignore($ident)
+
+Binds a meta variable for repetition, but expands to nothing.
+
+```rust
+macro_rules! count {
+    ( $( $i:stmt ),* ) => {{
+        0 $( + 1 ${ignore($i)} )*
+    }};
+}
+
+fn main() {
+    assert_eq!(count!(if true {} else {}, let _: () = (), || false), 3);
+}
+```
+
+Sometimes it is desired to repeat an expansion the same number of times as a meta variable repeats but without actually expanding the meta variable. It may be possible to work around this by expanding the meta variable in an expression like `{ let _ = $x; 1 }`, where the expanded value of `$x` is ignored, but this is only possible if what `$x` expands to is valid in this kind of expression.
+
+The `ignore` meta variable acts as if ident was used for the purposes of repetition, but expands to nothing.
+
+### index(depth)
+
+Expands to an unsuffixed integer literal representing the current iteration index of a ***repetition*** at the given depth.
+
+The output of `index` depends on where it is placed as well the provided index. If no index is provided, then it will always start at the innermost level.
+
+```rust
+trait Length {
+    fn length(&self) -> usize;
+}
+
+macro_rules! impl_tuple {
+    ( $( $name:ident ),* ) => {
+        impl<$( $name, )*> Length for ($( $name, )*)
+        where
+            $( $name: AsRef<[u8]>, )*
+        {
+            fn length(&self) -> usize {
+                let mut sum: usize = 0;
+                $({
+                    ${ignore($name)}
+                    sum = sum.wrapping_add(self.${index()}.as_ref().len());
+                })*
+                sum
+            }
+        }
+    };
+}
+
+impl_tuple!(A, B);
+
+fn main() {
+    assert_eq!((&[1, 2, 3][..], &[4, 5][..]).length(), 5);
+}
+```
+
+If repetitions are nested, then the optional depth parameter can be used to limit the number of nested repetitions that are counted.
+
+```rust
+macro_rules! innermost0 {
+    ( $( $a:ident: $( $b:literal ),* );+ ) => {
+        [$( $( ${ignore($b)} ${index()}, )* )+]
+    };
+}
+
+macro_rules! innermost1 {
+    ( $( $a:ident: $( $b:literal ),* );+ ) => {
+        [$( $( ${ignore($b)} ${index(1)}, )* )+]
+    };
+}
+
+macro_rules! outermost {
+    ( $( $a:ident: $( $b:literal ),* );+ ) => {
+        [$( ${ignore($a)} ${index()}, )+]
+    };
+}
+
+fn main() {
+    // 1 2 3 = 3 elements
+    // 4 5 = 2 elements
+    //
+    // Increasing list from the innermost loop refering innermost indexes
+    assert_eq!(innermost0!(a: 1, 2, 3; b: 4, 5), [0, 1, 2, 0, 1]);
+    
+    // a b = 2 elements
+    //
+    // Increasing list from the innermost loop refering outermost indexes
+    assert_eq!(innermost1!(a: 1, 2, 3; b: 4, 5), [0, 0, 0, 1, 1]);
+    
+    // a b = 2 elements
+    //
+    // Increasing list from the outermost loop refering outermost indexes
+    assert_eq!(outermost!(a: 1, 2, 3; b: 4, 5), [0, 1]);
+}
+```
+
+### length(depth)
```

Comment:
Bikeshedding, but should consider changing the name to "len". Length is abbreviated to "len" everywhere in `std`; there is only a single, unstable, API that uses "length".

Reply:
@jdahlstrom you may want to bring that up on the PR, currently under FCP proposal https://github.com/rust-lang/rust/pull/122808

Reply:
@tgross35 Indeed, done!